builder/amazon/common/state.go

```diff
@@ -35,228 +32,279 @@ type StateChangeConf struct {
+// poll delay and use the waiter-specific default
 
-// Returns 300 seconds (5 minutes) by default
-// Some AWS operations, like copying an AMI to a distant region, take a very long time
-// Allow user to override with AWS_TIMEOUT_SECONDS environment variable
-func TimeoutSeconds() (seconds int) {
-	seconds = 300
+// if AWS_TIMEOUT_SECONDS is set but AWS_MAX_ATTEMPTS is not, Packer will use
+// AWS_TIMEOUT_SECONDS and _either_ AWS_POLL_DELAY_SECONDS _or_ 2 if the user has not set AWS_POLL_DELAY_SECONDS, to determine a max number of attempts to make.
 
-	override := os.Getenv("AWS_TIMEOUT_SECONDS")
-	if override != "" {
-		n, err := strconv.Atoi(override)
-		if err != nil {
-			log.Printf("Invalid timeout seconds '%s', using default", override)
-		} else {
-			seconds = n
+// if AWS_TIMEOUT_SECONDS, _and_ AWS_MAX_ATTEMPTS are both set,
+// AWS_TIMEOUT_SECONDS will be ignored.
+
+// if AWS_MAX_ATTEMPTS is set but AWS_POLL_DELAY_SECONDS is not, then we will
+// use waiter-specific defaults.
+
+func getWaiterOptions() []request.WaiterOption {
+	waitOpts := make([]request.WaiterOption, 0)
+	// If user has set poll delay seconds, overwrite it. If user has NOT,
+	// default to a poll delay of 2 seconds
+	delayOverridden, delay := getEnvOverrides(2, "AWS_POLL_DELAY_SECONDS")
+	if delayOverridden {
+		delaySeconds := request.ConstantWaiterDelay(time.Duration(delay) * time.Second)
+		waitOpts = append(waitOpts, request.WithWaiterDelay(delaySeconds))
+	}
+
+	// If user has set max attempts, overwrite it. If user hasn't set max
+	// attempts, default to whatever the waiter has set as a default.
+	maxAttemptsOverridden, maxAttempts := getEnvOverrides(0, "AWS_MAX_ATTEMPTS")
+	if maxAttemptsOverridden {
+		waitOpts = append(waitOpts, request.WithWaiterMaxAttempts(maxAttempts))
+	}
+
+	timeoutOverridden, timeoutSeconds := getEnvOverrides(300, "AWS_TIMEOUT_SECONDS")
+	if maxAttemptsOverridden {
+		log.Printf("WARNING: AWS_MAX_ATTEMPTS and AWS_TIMEOUT_SECONDS are" +
+			" both set. Packer will be using AWS_MAX_ATTEMPTS and discarding " +
+			"AWS_TIMEOUT_SECONDS. If you have not set AWS_POLL_DELAY_SECONDS, " +
+			"Packer will default to a 2 second poll delay.")
+	} else if timeoutOverridden {
+		log.Printf("DEPRECATION WARNING: env var AWS_TIMEOUT_SECONDS is " +
+			"deprecated in favor of AWS_MAX_ATTEMPTS. If you have not " +
+			"explicitly set AWS_POLL_DELAY_SECONDS, we are defaulting to a " +
+			"poll delay of 2 seconds, regardless of the AWS waiter's default.")
+		maxAttempts := timeoutSeconds / delay
+		// override the delay so we can get the timeout right
+		if !delayOverridden {
+			delaySeconds := request.ConstantWaiterDelay(time.Duration(delay) * time.Second)
+			waitOpts = append(waitOpts, request.WithWaiterDelay(delaySeconds))
 		}
+		waitOpts = append(waitOpts, request.WithWaiterMaxAttempts(maxAttempts))
+	}
+	if len(waitOpts) == 0 {
+		log.Printf("No AWS timeout and polling overrides have been set. " +
```

Comment:
Ran a build with this patch and it looks like this line is output every time we check for the variables. I think they only need to be output once. Perhaps we could run this function in init, or store the output in the `state`/global?